src/udev/udev-watch.c

```diff
@@ -161,36 +165,450 @@ static int synthesize_change_one(sd_device *dev, sd_device *target) {
+        if (partitions)
+                log_device_debug(dev, "All partitions found by blkid are available on sysfs, triggering 'change' events without requesting the kernel to reread partition table.");
+        else
+                log_device_debug(dev, "No partition found by blkid and on sysfs, triggering 'change' event for the disk without requesting the kernel to reread partition table.");
+
+        (void) synthesize_change_all(dev, e);
+        return 1; /* synced and sent change events */
+}
+
+#define WAIT_SYNCED_DEFAULT_TIMEOUT_USEC (30 * USEC_PER_SEC)
+#define WAIT_SYNCED_INITIAL_DELAY_USEC   (100 * USEC_PER_MSEC)
+
+static usec_t sync_partition_timeout_usec(void) {
+        static usec_t saved = WAIT_SYNCED_DEFAULT_TIMEOUT_USEC;
+        static bool parsed = false;
+        int r;
+
+        if (parsed)
+                return saved;
+
+        parsed = true;
+
+        const char *e = getenv("SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC");
+        if (!e)
+                return saved;
+
+        usec_t timeout;
+        r = parse_sec(e, &timeout);
+        if (r < 0)
+                log_debug_errno(r, "Failed to parse $SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC=%s, ignoring: %m", e);
+
+        if (timeout > 5 * USEC_PER_HOUR) /* Add an arbitrary upper bound for safety */
+                log_debug("Parsed $SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC=%s is too large, ignoring.", e);
+        else
+                saved = timeout;
+
+        return saved;
+}
+
+typedef struct SyncDeviceContext {
+        sd_device *dev;
+        sd_event *event;
+        sd_event_source *timer;
+        usec_t delay;
+} SyncDeviceContext;
+
+static void sync_device_context_done(SyncDeviceContext *c) {
+        assert(c);
+
+        sd_device_unref(c->dev);
+        sd_event_unref(c->event);
+        sd_event_source_unref(c->timer);
+}
+
+static int on_watch_timer(sd_event_source *s, uint64_t usec, void *userdata);
+
+static int context_check_partition_synced(SyncDeviceContext *c) {
+        sd_device *dev = ASSERT_PTR(ASSERT_PTR(c)->dev);
+        int r;
+
+        r = partitions_synced(dev);
+        if (r == -EBUSY) {
+                r = event_reset_time_relative(
+                                c->event, &c->timer, CLOCK_MONOTONIC, c->delay, /* accuracy = */ 0,
+                                on_watch_timer, &c, SD_EVENT_PRIORITY_NORMAL,
+                                "watch-timer", /* force_reset = */ false);
+                if (r < 0)
+                        return log_device_debug_errno(dev, r, "Failed to reset timer: %m");
+
+                /* Let's exponentially increase the delay, to make not torture the disk. */
+                c->delay *= 2;
+                return 0;
+        }
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to check if partitions are synced: %m");
+        if (r == 0)
+                c->delay = WAIT_SYNCED_INITIAL_DELAY_USEC;
```

Comment:
I don't get why we need to reset the delay here.

Reply:
As the device is not locked now. When we receive an event later, we may be blocked because of the flock(), and in that case, we should not start the large delay.